### PR TITLE
refactor(es/ast): Improve type definitions of patterns

### LINKED ISSUE #8026: Concise design from oxc: `AssignmentTarget`
I'm filing this because I'm going to do this.

Concise design from oxc: `AssignmentTarget`

[OXC Playground](https://web-infra-dev.github.io/oxc/playground/?code=3YCAAIAGgICAgICAgICzm2zAvZAXeWWe4worViTTFBDYNTJqtWTSBHkPNuis7hwntGSGzihMEF5AqXc6Uo4rKgIQs2wkZFImzHf%2Ff1LjgIA%3D)
[SWC AST](https://swc-ast.vercel.app/#code/GYexAIF5wUwDwA4CcDcAoUIB0AjAhklLIqhmOANTTzLqa4GXUnpoDamAukTaR2AyTdmtNGSQAKTOBDBiyAJQBvAL7ipA-IVnykytUA)

https://github.com/web-infra-dev/oxc/blob/09ebe961d2a6b859f627af86f666d57d28b6a9b3/crates/oxc_ast/src/ast/js.rs#L746-L763

In swc, this structure is called `PatOrExpr`, while in oxc, it is referred to `AssignmentTarget`.

`AssignmentTarget` is an enum with two variants: `SimpleAssignmentTarget` and `AssignmentTargetPattern`.

`SimpleAssignmentTarget` wraps an identifier, or member expression (or TypeScript expression).
`AssignmentTargetPattern` corresponds to `Pat` in swc.

This design offers the following advantages:

1. Explicit lhs. By marking the lhs during the parsing phase, we can easily determine whether an expression is an lhs during the transformation process. With the presence of `AssignmentTarget`, the lhs can only be an `AssignmentTarget`.
2. Non-overlapping structures. When an identifier appears within `AssignmentTarget`, it can only be under `SimpleAssignmentTarget`. (However, in TypeScript, there is an additional layer of wrapping, which can be disregarded if the TS pass has already been executed.)
3. Convenient conversion. `AssignmentTarget` is a limited expression, not arbitrary. This is particularly useful when converting between `foo` and `foo.bar`, as there is no need to cross the boundaries of `Pat` and `Expr`.

_Originally posted by @magic-akari in https://github.com/swc-project/swc/discussions/5186#discussioncomment-7118889_

### PR DESCRIPTION
**Pull request by kdy1**
**Description:**

 - Copy `AssignmentTarget` from `oxc`.
 - Use `BindingIdentifier` in more places.

**Related issue:**
 
 - Closes #8026

### CODE REVIEW

**Comment by Boshen** [comment]
The AST is from [jsparagus](https://github.com/mozilla-spidermonkey/jsparagus), all credits to them!

**Review by kdy1 — COMMENTED** [review]
swc-bump:
 - swc_ecma_ast --breaking

**Comment by kdy1** [comment]
Status update: I'm still focusing on this issue, and there's still enormous amount of code to fix 
image

File: crates/swc_ecma_transforms_typescript/src/transform.rs [hunk 1]
```diff
@@ -232,7 +232,8 @@ impl VisitMut for Transform {
                 .flat_map(|key| {
                     key.init.take().map(|init| {
                         let name = key.name.clone();
-                        init.make_assign_to(op!("="), name.into())
+                        dbg!(&name);
```
Comment (kdy1) [inline crates/swc_ecma_transforms_typescript/src/transform.rs:235]:
```suggestion
```

**Pull request closed without merging**